sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala

```diff
@@ -389,6 +389,102 @@ case class AppendColumnsWithObjectExec(
+ */
+case class MapGroupsWithInitialStateExec(
+    func: (Any, Iterator[Any], LogicalGroupState[Any]) => TraversableOnce[Any],
+    keyDeserializer: Expression,
+    valueDeserializer: Expression,
+    initialStateDeserializer: Expression,
+    groupingAttributes: Seq[Attribute],
+    initialStateGroupingAttributes: Seq[Attribute],
+    dataAttributes: Seq[Attribute],
+    initialStateDataAttrs: Seq[Attribute],
+    outputObjAttr: Attribute,
+    initialState: SparkPlan,
+    timeoutConf: GroupStateTimeout,
+    child: SparkPlan) extends BinaryExecNode with ObjectProducerExec {
+
+  override def left: SparkPlan = child
+
+  override def right: SparkPlan = initialState
+
+  override def outputPartitioning: Partitioning = child.outputPartitioning
+
+  private val watermarkPresent = child.output.exists {
+    case a: Attribute if a.metadata.contains(EventTimeWatermark.delayKey) => true
+    case _ => false
+  }
+
+  /**
+   * Distribute by grouping attributes - We need the underlying data and the initial state data
+   * to have the same grouping so that the data are co-lacated on the same task.
+   */
+  override def requiredChildDistribution: Seq[Distribution] = {
+    ClusteredDistribution(groupingAttributes) ::
+      ClusteredDistribution(initialStateGroupingAttributes) :: Nil
+  }
+
+  /**
+   * Ordering needed for using GroupingIterator.
+   * We need the initial state to also use the ordering as the data so that we can co-locate the
+   * keys from the underlying data and the initial state.
+   */
+  override def requiredChildOrdering: Seq[Seq[SortOrder]] = Seq(
+    groupingAttributes.map(SortOrder(_, Ascending)),
+    initialStateGroupingAttributes.map(SortOrder(_, Ascending)))
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    child.execute().zipPartitions(
+      initialState.execute()
+    ) { case (dataIter, initialStateIter) =>
+      val groupedChildDataIter = GroupedIterator(dataIter, groupingAttributes, child.output)
+      val groupedInitialStateIter = GroupedIterator(
+        initialStateIter, initialStateGroupingAttributes, initialState.output)
+      val getKey = ObjectOperator.deserializeRowToObject(keyDeserializer, groupingAttributes)
+      val getValue = ObjectOperator.deserializeRowToObject(valueDeserializer, dataAttributes)
+      val outputObject = ObjectOperator.wrapObjectToRow(outputObjectType)
+      val getStateObj =
+        ObjectOperator.deserializeRowToObject(initialStateDeserializer, initialStateDataAttrs)
+
+      new CoGroupedIterator(
+        groupedChildDataIter, groupedInitialStateIter, groupingAttributes).flatMap {
+        case (keyRow, valueRowIter, initialStateRowIter) =>
+          var foundInitialStateForKey = false
+          val optionalState = initialStateRowIter.map { initialStateRow =>
+            if (foundInitialStateForKey) {
+              throw new IllegalArgumentException("The initial state provided contained " +
```

Comment:
This error should be made into a common error between FlatMapGroupswithStateExec and this class. If possible.